aptos-move/framework/aptos-stdlib/sources/cryptography/groups.move

```diff
@@ -0,0 +1,749 @@
+        is_prime_order_internal<G>()
+    }
+
+    /// Get the order of group `G`, little-endian encoded as a byte string.
+    public fun group_order<G>(): vector<u8> {
+        abort_if_feature_disabled();
+        group_order_internal<G>()
+    }
+
+    #[test_only]
+    /// Generate a random group element.
+    public fun random_element<G>(): Element<G> {
+        abort_if_feature_disabled();
+        Element<G> {
+            handle: random_element_internal<G>()
+        }
+    }
+
+    #[test_only]
+    /// Generate a random scalar.
+    public fun random_scalar<G>(): Scalar<G> {
+        abort_if_feature_disabled();
+        Scalar<G> {
+            handle: random_scalar_internal<G>()
+        }
+    }
+
+    fun abort_if_feature_disabled() {
+        if (!std::features::generic_groups_enabled()) {
+            abort(std::error::invalid_state(E_NATIVE_FUN_NOT_AVAILABLE))
+        };
+    }
+
+    // Native functions.
+    native fun deserialize_element_uncompressed_internal<G>(bytes: vector<u8>): (bool, u64);
+    native fun deserialize_element_compressed_internal<G>(bytes: vector<u8>): (bool, u64);
+    native fun scalar_from_u64_internal<G>(value: u64): u64;
+    native fun deserialize_scalar_internal<G>(bytes: vector<u8>): (bool, u64);
+    native fun scalar_neg_internal<G>(handle: u64): u64;
+    native fun scalar_add_internal<G>(handle_1: u64, handle_2: u64): u64;
+    native fun scalar_double_internal<G>(handle: u64): u64;
+    native fun scalar_mul_internal<G>(handle_1: u64, handle_2: u64): u64;
+    native fun scalar_inv_internal<G>(handle: u64): (bool, u64);
+    native fun scalar_eq_internal<G>(handle_1: u64, handle_2: u64): bool;
+    native fun serialize_scalar_internal<G>(h: u64): vector<u8>;
+    native fun element_add_internal<G>(handle_1: u64, handle_2: u64): u64;
+    native fun element_eq_internal<G>(handle_1: u64, handle_2: u64): bool;
+    native fun group_identity_internal<G>(): u64;
+    native fun is_prime_order_internal<G>(): bool;
+    native fun group_order_internal<G>(): vector<u8>;
+    native fun group_generator_internal<G>(): u64;
+    native fun element_mul_internal<G>(scalar_handle: u64, element_handle: u64): u64;
+    native fun element_double_internal<G>(element_handle: u64): u64;
+    native fun element_neg_internal<G>(handle: u64): u64;
+    native fun serialize_element_uncompressed_internal<G>(handle: u64): vector<u8>;
+    native fun serialize_element_compressed_internal<G>(handle: u64): vector<u8>;
+    native fun element_multi_scalar_mul_internal<G>(scalar_handles: vector<u64>, element_handles: vector<u64>): u64;
+    native fun pairing_product_internal<G1,G2,Gt>(g1_handles: vector<u64>, g2_handles: vector<u64>): u64;
+    native fun hash_to_element_internal<G>(bytes: vector<u8>): u64;
+    #[test_only]
+    native fun random_element_internal<G>(): u64;
+    #[test_only]
+    native fun random_scalar_internal<G>(): u64;
+
+    #[test(fx = @std)]
+    fun test_bls12_381_g1(fx: signer) {
```

Comment:
These tests are great!

I think eventually, we'll want to add generic tests that take the group `G` as a type argument and test things randomly.

e.g., you can test scalar multiplication by picking a random base `H` and two random scalars `a` and `b` and checking if `(a + b) * H = a * H + b * H`

There will be some things we cannot test like this though (e.g., serialization).

Reply:
Fixed.